src/index.ts

```diff
@@ -1,11 +1,202 @@
+      index: indexFromLogGroup(data.logGroup),
+      event: event.message,
+      fields: {
+        account: envVars.account,
+        environment: envVars.environment
+      }
+    }
+  })
+}
+
+function transformALBLog(data: S3LogRecord, envVars: EnvVars): SplunkRecord[] {
+  return data.Logs.map((log) => {
+    return {
+      host: data.ALB as string,
+      source: 'ALB',
+      sourcetype: 'aws:elb:accesslogs',
+      index: 'pay_ingress',
+      event: log,
+      fields: {
+        account: envVars.account,
+        environment: envVars.environment
+      }
+    }
+  })
+}
+
+function transformS3AccessLog(data: S3LogRecord, envVars: EnvVars): SplunkRecord[] {
+  return data.Logs.map((log) => {
+    return {
+      host: data.S3Bucket as string,
+      source: 'S3',
+      sourcetype: 'aws:s3:accesslogs',
+      index: 'pay_access',
+      event: log,
+      fields: {
+        account: envVars.account,
+        environment: envVars.environment
+      }
+    }
+  })
+}
+
+function shouldDropRecord(data: object): boolean {
+  if('messageType' in data && data.messageType !== 'DATA_MESSAGE') {
+    return true
+  }
+  return false
+}
+
+function transformData(data: object, envVars: EnvVars): SplunkRecord[] {
+  if( 'logGroup' in data) {
+    return transformCloudWatchData(data as CloudWatchRecordData, envVars)
+  } else if ('ALB' in data) {
+    return transformALBLog(data as S3LogRecord, envVars)
+  } else if ('S3Bucket' in data) {
+    return transformS3AccessLog(data as S3LogRecord, envVars)
+  }
+  throw new Error('Cannot parse information from record data because it is an unregonised structure.')
+}
+
+function getEnvVar(varName: string): string {
+  const value = process.env[varName]
+  if (!value) {
+    throw new Error(`"${varName}" env var is not set`)
+  }
+  return value
+}
+
+function getEnvVars(): EnvVars {
+  return {
+    environment: getEnvVar('ENVIRONMENT'),
+    account: getEnvVar('ACCOUNT')
+  }
+}
+
+export const handler: Handler = async (event: FirehoseTransformationEvent): Promise<FirehoseTransformationResult> => {
+  const envVars = getEnvVars()
+
+  const records: FirehoseTransformationResultRecord[] = []
+  for (const record of event.records) {
+    const recordData = JSON.parse(Buffer.from(record.data, 'base64').toString())
+
+    if (shouldDropRecord(recordData)) {
+      records.push({
+        recordId: record.recordId,
+        result: 'Dropped',
+        data: record.data
+      })
+    } else {
+      try{
+        const transformedData = transformData(recordData, envVars)
+        const joinedData = transformedData.map(x=> JSON.stringify(x)).join('\n')
+        records.push({
+          recordId: record.recordId,
+          result: 'Ok',
+          data: Buffer.from(joinedData).toString('base64')
+        })
+      } catch(e) {
+        throw new Error(`Error processing record "${record.recordId}": ${(e as Error).message}`)
+      }
+    }
+  }
+
+//  //TODO: remove this debugging when its ready
```

Comment:
Would be useful to add some logging to debug lambda for any issues. similar to https://github.com/alphagov/pay-logging-s3-to-firehose-delivery/blob/main/src/index.ts#L187

Reply:
Yep, nice idea I'll put that debug in a function and use a DEBUG env var toggle 👍